### PR TITLE
JAT-30 Support Loading and Saving presets

### PR DESCRIPTION
**Pull request by flyingpandaboy**
This includes fetching presets folder list.

### CODE REVIEW

**Comment by flyingpandaboy** [comment]
I think we should create a presets/ folder on startup so that fetching, saving won't bug out. I'd prefer on startup than in the functions as we don't know what order the functions would execute in. It'd probably be in fetching though but that seems weird to make a folder in there.

**Comment by flyingpandaboy** [comment]
Only thing pending is to use getPresetListFromFiles() to update our UI

File: src/main/flush.ts [hunk 1]
```diff
@@ -64,6 +73,32 @@ export const save = (state: IState) => {
...
+    });
+    return JSON.parse(content) as IPreset;
+  } catch (ex) {
+    console.log('Failed to get presets!!');
+    throw ex;
+  }
+};
+
+export const savePreset = (preset_name: string, preset_info: IPreset) => {
+  // need to create presets dir first...... should we do this in the install?
```
Comment (selinali2010) [inline src/main/flush.ts:89]: I think this comment isn't necessary anymore now that we're creating a presets dir right? Let's remove it if we can.

File: src/main/main.ts [hunk 2]
```diff
@@ -148,6 +159,49 @@ ipcMain.on(ChannelEnum.HEALTH_CHECK, async (event) => {
   }
 });
 
+ipcMain.on(ChannelEnum.LOAD_PRESET, async (event, preset_name: string) => {
+  const channel = ChannelEnum.LOAD_PRESET;
+  console.log(`Loading preset: ${preset_name}`);
+
+  // TODO: should we do some str checking here?
+  // might have to make the preset folder too?
```
Comment (selinali2010) [inline src/main/main.ts:167]: We can clean up these comments too. We don't seem to be doing str checking per se but we ARE throwing an error if we fail to fetch so that seems good enough to me.

File: src/main/main.ts [hunk 3]
```diff
@@ -148,6 +159,49 @@ ipcMain.on(ChannelEnum.HEALTH_CHECK, async (event) => {
...
+    filters: state.filters,
+  });
+  await handleUpdate(event, channel);
+});
+
+ipcMain.on(ChannelEnum.GET_PRESET_FILE_LIST, async (event) => {
+  const channel = ChannelEnum.GET_PRESET_FILE_LIST;
+  console.log(`Getting Preset List`);
+
+  // const fileNames: string[] = [];
```
Comment (selinali2010) [inline src/main/main.ts:194]: unused code?

File: src/renderer/PresetsBar.tsx [hunk 4]
```diff
@@ -20,16 +21,16 @@ const PresetListItem = ({
   handleDelete,
   isDisabled,
 }: IListItemProps) => {
-  const editValueRef = useRef<HTMLInputElement>(null);
+  const editModeRef = useRef<HTMLInputElement>(null);
```
Comment (selinali2010) [inline src/renderer/PresetsBar.tsx:24]: Can we undo these changes from `editValueRef` to `editModeRef`? This was a change I made in jat-90 based on @xenown's feedback

File: src/renderer/PresetsBar.tsx [hunk 5]
```diff
@@ -113,29 +114,37 @@ const PresetsBar = () => {
...
-      // TODO: overwrite preset values for the existing preset
-    } else {
-      // TODO: add code for creating a preset
-      setSelectedPresetName(presetName);
+  const handleCreatePreset = async (prev: string[]) => {
+    await savePreset(presetName);
+
+    // If we are creating a new preset and not just updating an existing one
+    if (prev.indexOf(presetName) === -1) {
+      await setSelectedPresetName(presetName);
```
Comment (selinali2010) [inline src/renderer/PresetsBar.tsx:122]: Since `setSelectedPresetName` doesn't return a `Promise`, the `await` keyword isn't necessary here. The existence of the `await` in line 188 means that the code below it will be treated as if it were wrapped in a `.then(() => { ... })`.

File: src/renderer/PresetsBar.tsx [hunk 6]
```diff
@@ -113,29 +114,37 @@ const PresetsBar = () => {
...
       const newPresets = [...prev, presetName].sort();
-      setPresetNames(newPresets);
+      await setPresetNames(newPresets);
```
Comment (selinali2010) [inline src/renderer/PresetsBar.tsx:124]: same here, can remove `await`

File: src/renderer/PresetsBar.tsx [hunk 7]
```diff
@@ -113,29 +114,37 @@ const PresetsBar = () => {
...
+    console.log('Finished Handling Create Preset!');
   };
 
-  const handleLoadPreset = () => {
-    // TODO: add code for loading a preset
+  const handleLoadPreset = async () => {
+    if (selectedPresetName) {
+      try {
+        await loadPreset(selectedPresetName);
+        await performHealthCheck();
```
Comment (selinali2010) [inline src/renderer/PresetsBar.tsx:134]: same here, can remove `await`

File: src/renderer/PresetsBar.tsx [hunk 8]
```diff
@@ -113,29 +114,37 @@ const PresetsBar = () => {
...
   const handleChangeNewPresetName = (newValue: string) => {
     setPresetName(newValue);
 
-    // Update selectedPresetName if the input value matches an existing preset name
-    const newSelectedName = presetNames.some((n) => n === newValue)
-      ? newValue
-      : undefined;
-    setSelectedPresetName(newSelectedName);
+    // TODO: If the preset name is an existing one, we should set the selected one to be that so we can load it
+    setSelectedPresetName(undefined);
```
Comment (selinali2010) [inline src/renderer/PresetsBar.tsx:147]: Could we undo this change? It was a fix I made in jat-90 that I force-pushed so it likely didn't make it into this branch.

**Review by selinali2010 — COMMENTED** [review]
Some small nits from me!

File: src/main/main.ts [hunk 9]
```diff
@@ -148,6 +159,45 @@ ipcMain.on(ChannelEnum.HEALTH_CHECK, async (event) => {
   }
 });
 
+ipcMain.on(ChannelEnum.LOAD_PRESET, async (event, preset_name: string) => {
+  const channel = ChannelEnum.LOAD_PRESET;
+  console.log(`Loading preset: ${preset_name}`);
+
+  // TODO: should we do some str checking here?
```
Comment (selinali2010) [inline src/main/main.ts:166]: While we aren't doing strict str checking, we're throwing an error if we fail to fetch so that seems good enough to me. @xenown what do you think about this?

File: src/main/flush.ts [hunk 10]
```diff
@@ -1,6 +1,11 @@
...
 import path from 'path';
-import { getDefaultState, IState } from '../common/constants';
+import {
+  getDefaultState,
+  IPreset,
+  IState,
+  PRESETS_DIR,
+} from '../common/constants';
 
 export const stateToString = (state: IState) => {
```
Comment (xenown) [inline src/main/flush.ts:10]: maybe we should call this serializeState?

File: src/main/flush.ts [hunk 11]
```diff
@@ -64,6 +73,31 @@ export const save = (state: IState) => {
   }
 };
 
+export const fetchPreset = (preset_name: string) => {
+  try {
+    const content = fs.readFileSync(PRESETS_DIR + preset_name, {
```
Comment (xenown) [inline src/main/flush.ts:78]: Please use addFileToPath() or path.join

File: src/main/flush.ts [hunk 12]
```diff
@@ -64,6 +73,31 @@ export const save = (state: IState) => {
...
+    return JSON.parse(content) as IPreset;
+  } catch (ex) {
+    console.log('Failed to get presets!!');
+    throw ex;
+  }
+};
+
+export const savePreset = (preset_name: string, preset_info: IPreset) => {
+  try {
+    fs.writeFileSync(PRESETS_DIR + preset_name, serializePreset(preset_info), {
```
Comment (xenown) [inline src/main/flush.ts:90]: same comment as above.

File: src/main/main.ts [hunk 13]
```diff
@@ -74,6 +80,11 @@ const setWindowDimension = (isExpanded: boolean) => {
 const state: IState = fetchSettings();
 let configPath = '';
 
+// create presets dir if it doesn't exist
+if (!fs.existsSync(PRESETS_DIR)) {
```
Comment (xenown) [inline src/main/main.ts:84]: The node docs aren't clear but I think fs functions will throw exceptions. https://stackoverflow.com/questions/60497053/exception-in-node-when-using-mkdirsync also suggests try catch. Can we catch exceptions here?

File: src/main/main.ts (hunk 9 above)
Comment (xenown) [reply src/main/main.ts:180]: I think its fine. 177 and 178 won't throw exceptions.

File: src/main/main.ts [hunk 14]
```diff
@@ -162,6 +173,45 @@ ipcMain.on(ChannelEnum.HEALTH_CHECK, async (event) => {
...
+    filters: state.filters,
+  });
+  await handleUpdate(event, channel);
+});
+
+ipcMain.on(ChannelEnum.GET_PRESET_FILE_LIST, async (event) => {
+  const channel = ChannelEnum.GET_PRESET_FILE_LIST;
+  console.log(`Getting Preset List`);
+
+  const fileNames: string[] = fs.readdirSync(PRESETS_DIR);
```
Comment (xenown) [inline src/main/main.ts:206]: Same thing, catch exception here

File: src/renderer/PresetsBar.tsx [hunk 15]
```diff
@@ -84,48 +90,51 @@ const PresetListItem = ({
...
-    if (prev.indexOf(presetName) !== -1) {
-      // TODO: overwrite preset values for the existing preset
-    } else {
-      // TODO: add code for creating a preset
+  const [presetNames, setPresetNames] = useState<string[]>([]);
+
+  // Fetch default presets and custom presets from storage
+  useEffect(() => {
+    const fetchPresetNames = async () => {
+      const result = await getPresetListFromFiles();
```
Comment (xenown) [inline src/renderer/PresetsBar.tsx:105]: Need to handle exception case

File: src/renderer/PresetsBar.tsx [hunk 16]
```diff
@@ -84,48 +90,51 @@ const PresetListItem = ({
...
+      // Sort preset names before updating state just in case
+      setPresetNames(result.sort());
+    };
+
+    fetchPresetNames();
+  }, []);
+
+  const handleCreatePreset = async (prev: string[]) => {
+    await savePreset(presetName);
```
Comment (xenown) [inline src/renderer/PresetsBar.tsx:114]: exception case

File: src/renderer/PresetsBar.tsx [hunk 17]
```diff
@@ -84,48 +90,51 @@ const PresetListItem = ({
...
+
+    // If we are creating a new preset and not just updating an existing one
+    if (prev.indexOf(presetName) === -1) {
       setSelectedPresetName(presetName);
+      // Keep presets sorted
       const newPresets = [...prev, presetName].sort();
       setPresetNames(newPresets);
     }
+
+    console.log('Finished Handling Create Preset!');
```
Comment (xenown) [inline src/renderer/PresetsBar.tsx:124]: Remove this once you're done debugging.

File: src/renderer/PresetsBar.tsx [hunk 18]
```diff
@@ -84,48 +90,51 @@ const PresetListItem = ({
...
 
-  const handleLoadPreset = () => {
-    // TODO: add code for loading a preset
+  const handleLoadPreset = async () => {
+    if (selectedPresetName) {
+      try {
+        await loadPreset(selectedPresetName);
+        performHealthCheck();
+      } catch (e) {
+        console.log(`failed to get preset ${selectedPresetName}`);
```
Comment (xenown) [inline src/renderer/PresetsBar.tsx:133]: Remove console log

File: src/renderer/SideBar.tsx [hunk 19]
```diff
@@ -16,7 +16,7 @@ const SideBar = () => {
 
   const setGain = async (newValue: number) => {
     try {
-      await setAutoPreAmpOn(false);
+      setAutoPreAmpOn(false);
       await setMainPreAmp(newValue);
```
Comment (xenown) [inline src/renderer/SideBar.tsx:20]: We forgot to tell backend to update autopreamp setting when we first wrote this file. lets fix this.

**Review by xenown — CHANGES_REQUESTED** [review]
Mostly need error catching and remove console logs. Also theres still the edit preset name to existing name case to impl.

File: src/main/main.ts (hunk 13 above)
Comment (selinali2010) [reply src/main/main.ts:84]: Just to clarify, are you referring to the `fs.existsSync` function or the `fs.mkdirSync` function? (or both...?)

File: src/main/flush.ts (hunk 10 above)
Comment (flyingpandaboy) [reply src/main/flush.ts:10]: sure, we can but think we should do another PR for that.

File: src/main/main.ts (hunk 13 above)
Comment (flyingpandaboy) [reply src/main/main.ts:84]: Sure, we can't preset if presets folder fails to create though. Added try catch but still throw e if it happens

File: src/main/main.ts (hunk 14 above)
Comment (flyingpandaboy) [reply src/main/main.ts:213]: ok, i'll send 500 back if it fails

File: src/renderer/PresetsBar.tsx (hunk 15 above)
Comment (flyingpandaboy) [reply src/renderer/PresetsBar.tsx:107]: right now just populating with empty list. We could do a popup or something? not sure how the error handling should be handled here

File: src/renderer/PresetsBar.tsx (hunk 16 above)
Comment (flyingpandaboy) [reply src/renderer/PresetsBar.tsx:116]: same, not sure what to do here. just continue running ig?

File: src/renderer/SideBar.tsx (hunk 19 above)
Comment (flyingpandaboy) [reply src/renderer/SideBar.tsx:20]: wdym by this?

File: src/renderer/PresetsBar.tsx (hunk 15 above)
Comment (selinali2010) [reply src/renderer/PresetsBar.tsx]: @xenown mentioned in the gc already, but you can set the global error by calling `setGlobalError(e);` - this will open up the `PrereqMissingModal` (which, in retrospect, we should really rename it to something more general like `ErrorModal`) that prompts users to either retry (aka refresh the page) or exit the app.

File: src/renderer/SideBar.tsx (hunk 19 above)
Comment (selinali2010) [reply src/renderer/SideBar.tsx:20]: The expected behavior is that we want to save the state of `isAutoPreAmpOn` in the BE whenever we change it. This is an instance of us changing the state in the UI but not updating the BE. We need to add:

```

File: src/renderer/PresetsBar.tsx [hunk 20]
```diff
@@ -102,26 +104,35 @@ const PresetsBar = () => {
...
-      // Sort preset names before updating state just in case
-      setPresetNames(result.sort());
+      try {
+        const result = await getPresetListFromFiles();
+        // Sort preset names before updating state just in case
+        setPresetNames(result.sort());
+      } catch (e) {
+        console.error('Failed to get preset list of files');
+        console.error(e);
+        setPresetNames(['']);
```
Comment (selinali2010) [inline src/renderer/PresetsBar.tsx:114]: Just curious, is this line necessary? `presetNames` is initialized as `[]` - is there an explicit reason why we want to give it an element when fetching fails?

File: src/renderer/PresetsBar.tsx [hunk 21]
```diff
@@ -44,6 +45,7 @@ const PresetListItem = ({
   };
 
   const handleDeleteClicked = () => {
+    deletePreset(value);
```
Comment (selinali2010) [inline src/renderer/PresetsBar.tsx:48]: could we move this into `handleDeletePreset` in the outer component and add an `await` + `try/catch` around `deletePreset`?

File: src/renderer/utils/equalizerApi.ts [hunk 22]
```diff
@@ -117,6 +117,16 @@ export const savePreset = (preset_name: string): Promise<void> => {
   return promisifyResult(setterResponseHandler, channel);
 };
 
+/**
+ * Delete a preset file in preset folder
+ * @returns { Promise<void> } if delete was succesfull
+ */
+export const deletePreset = (preset_name: string): Promise<void> => {
```
Comment (selinali2010) [inline src/renderer/utils/equalizerApi.ts:124]: sorry, minor thing, we've been using camel case for variable names for the most part - could we try to be more consistent with that?

File: src/renderer/PresetsBar.tsx (hunk 20 above)
Comment (flyingpandaboy) [reply src/renderer/PresetsBar.tsx:114]: changed to set global error

File: src/renderer/utils/equalizerApi.ts (hunk 22 above)
Comment (flyingpandaboy) [reply src/renderer/utils/equalizerApi.ts:124]: fixed

File: src/renderer/PresetsBar.tsx (hunk 21 above)
Comment (flyingpandaboy) [reply src/renderer/PresetsBar.tsx:48]: yeah

File: src/renderer/SideBar.tsx (hunk 19 above)
Comment (flyingpandaboy) [reply src/renderer/SideBar.tsx:20]: added

**Review by selinali2010 — APPROVED** [review]
Lgtm! @xenown plz review too

File: src/main/main.ts [hunk 23]
```diff
@@ -162,6 +179,60 @@ ipcMain.on(ChannelEnum.HEALTH_CHECK, async (event) => {
...
+    console.log('Failed to read preset: ', presetName);
+    throw ex;
+  }
+});
+
+ipcMain.on(ChannelEnum.SAVE_PRESET, async (event, arg) => {
+  const channel = ChannelEnum.SAVE_PRESET;
+  const presetName = arg[0];
+  console.log(`Saving preset: ${presetName}`);
+  savePreset(presetName, {
```
Comment (xenown) [inline src/main/main.ts:203]: any reason why we're not rethrowing the exception here?

File: src/main/main.ts [hunk 24]
```diff
@@ -162,6 +179,60 @@ ipcMain.on(ChannelEnum.HEALTH_CHECK, async (event) => {
...
+
+  // TODO: should we do some str checking here?
+  try {
+    const presetSettings: IPreset = fetchPreset(presetName);
+    state.preAmp = presetSettings.preAmp;
+    state.filters = presetSettings.filters;
+    await handleUpdate(event, channel);
+  } catch (ex) {
+    console.log('Failed to read preset: ', presetName);
+    throw ex;
```
Comment (xenown) [inline src/main/main.ts:195]: I'm not sure if I mentioned to throw exceptions but if I did, sorry thats not the right thing to do here. Instead, we should be calling handleError() and passing a corresponding return code. Add a corresponding message to errors.ts and then this should be good. See handleUpdateHelper for an example.

File: src/main/main.ts (hunk 23 above)
Comment (xenown) [inline src/main/main.ts:203]: handleError here too

File: src/main/main.ts [hunk 25]
```diff
@@ -162,6 +179,60 @@ ipcMain.on(ChannelEnum.HEALTH_CHECK, async (event) => {
...
+
+  try {
+    const fileNames: string[] = fs.readdirSync(PRESETS_DIR);
+    console.log(fileNames);
+    const reply: TSuccess<string[]> = { result: fileNames };
+    event.reply(channel, reply);
+  } catch (e) {
+    console.error('Failed to get filenames');
+    console.error(e);
+    const reply: TError = { errorCode: 500 };
```
Comment (xenown) [inline src/main/main.ts:231]: use HandleError

File: src/main/main.ts [hunk 26]
```diff
@@ -162,6 +179,60 @@ ipcMain.on(ChannelEnum.HEALTH_CHECK, async (event) => {
...
+  });
+  await handleUpdate(event, channel);
+});
+
+ipcMain.on(ChannelEnum.DELETE_PRESET, async (event, arg) => {
+  const channel = ChannelEnum.DELETE_PRESET;
+  const presetName = arg[0];
+  const pathToDelete = path.join(PRESETS_DIR, presetName);
+  console.log(`Deleting preset: ${presetName} at location ${pathToDelete}`);
+  fs.unlinkSync(pathToDelete);
```
Comment (xenown) [inline src/main/main.ts:215]: try catch this? and handleError()

File: src/renderer/PresetsBar.tsx [hunk 27]
```diff
@@ -84,48 +91,58 @@ const PresetListItem = ({
...
+
+      // If we are creating a new preset and not just updating an existing one
+      if (prev.indexOf(presetName) === -1) {
+        setSelectedPresetName(presetName);
+        // Keep presets sorted
+        const newPresets = [...prev, presetName].sort();
+        setPresetNames(newPresets);
+      }
+    } catch (e) {
+      console.error('Failed to save preset');
```
Comment (xenown) [inline src/renderer/PresetsBar.tsx:132]: Lets not keep console log statements?

File: src/renderer/PresetsBar.tsx [hunk 28]
```diff
@@ -148,12 +165,22 @@ const PresetsBar = () => {
...
+      );
     };
 
-    const handleDeletePreset = (deletedValue: string) => {
+    const handleDeletePreset = async (deletedValue: string) => {
       setPresetNames(presetNames.filter((n) => n !== deletedValue));
+      try {
+        await deletePreset(deletedValue);
+      } catch (e) {
+        console.log('Failed to delete preset!');
```
Comment (xenown) [inline src/renderer/PresetsBar.tsx:181]: no console.log

File: src/renderer/PresetsBar.tsx [hunk 29]
```diff
@@ -84,48 +91,58 @@ const PresetListItem = ({
...
+
+  // Fetch default presets and custom presets from storage
+  useEffect(() => {
+    const fetchPresetNames = async () => {
+      try {
+        const result = await getPresetListFromFiles();
+        // Sort preset names before updating state just in case
+        setPresetNames(result.sort());
+      } catch (e) {
+        console.error('Failed to get preset list of files');
```
Comment (xenown) [inline src/renderer/PresetsBar.tsx:111]: no log

**Review by xenown — CHANGES_REQUESTED** [review]

File: src/main/main.ts [hunk 30]
```diff
@@ -162,6 +178,70 @@ ipcMain.on(ChannelEnum.HEALTH_CHECK, async (event) => {
...
+  const pathToDelete = path.join(PRESETS_DIR, presetName);
+  console.log(`Deleting preset: ${presetName} at location ${pathToDelete}`);
+  try {
+    fs.unlinkSync(pathToDelete);
+  } catch (e) {
+    console.log('Failed to delete preset');
+    console.log(e);
+    handleError(event, channel, ErrorCode.PRESET_FILE_ERROR);
+  }
+  await handleUpdate(event, channel);
```
Comment (xenown) [inline src/main/main.ts:226]: lets move handleUpdate to the try block

File: src/main/main.ts [hunk 31]
```diff
@@ -74,6 +79,17 @@ const setWindowDimension = (isExpanded: boolean) => {
...
 let configPath = '';
 
+try {
+  // create presets dir if it doesn't exist
+  if (!fs.existsSync(PRESETS_DIR)) {
+    fs.mkdirSync(PRESETS_DIR);
+  }
+} catch (e) {
+  console.error('Failed to make presets directory!!');
+  console.error(e);
```
Comment (xenown) [inline src/main/main.ts:89]: idk if throwing an error is useful here. since this is tied to the scope of the app launch, i don't think this exception does anything. Technically, all our other preset functions will gracefully throw errors should this preset folder creation fail. Its fine to leave it like this. @flyingpandaboy @selinali2010 what do yall think?

File: src/main/main.ts (hunk 30 above)
Comment (xenown) [inline src/main/main.ts:226]: move to try block

**Review by xenown — CHANGES_REQUESTED** [review]

File: src/main/main.ts (hunk 31 above)
Comment (flyingpandaboy) [reply src/main/main.ts:89]: what do you mean by exception doesn't do anything? Won't the application crash on startup

File: src/common/errors.ts [hunk 32]
```diff
@@ -41,6 +42,12 @@ export const errors: Record<ErrorCode, ErrorDescription> = {
       'Please restart the application. If the error persists, try reaching out to the developers to resolve the issue.',
     code: ErrorCode.FAILURE,
   },
+  [ErrorCode.PRESET_FILE_ERROR]: {
+    shortError: 'Internal Error: Failed to read or modify preset files.',
+    action:
+      'Please check that the preset name is saveable as a file and that the installation directory is in a writeable place. In addition, check that you have available storage space. If the error persists, try reaching out to the developers to resolve the issue.',
+    code: ErrorCode.FAILURE,
```
Comment (selinali2010) [inline src/common/errors.ts:49]: should this be `code: ErrorCode.PRESET_FILE_ERROR`?

File: src/main/main.ts (hunk 31 above)
Comment (xenown) [reply src/main/main.ts:89]: Okay confirmed app crashes with the exception thrown.

File: src/main/main.ts [hunk 33]
```diff
@@ -162,6 +178,70 @@ ipcMain.on(ChannelEnum.HEALTH_CHECK, async (event) => {
...
+  console.log(`Saving preset: ${presetName}`);
+  try {
+    savePreset(presetName, {
+      preAmp: state.preAmp,
+      filters: state.filters,
+    });
+  } catch (e) {
+    handleError(event, channel, ErrorCode.PRESET_FILE_ERROR);
+  }
+  await handleUpdate(event, channel);
```
Comment (xenown) [inline src/main/main.ts:211]: move this too

**Review by xenown — CHANGES_REQUESTED** [review]

**Review by xenown — APPROVED** [review]
